### PR TITLE
Fix #2953: NoopAuthorizer should create a anonymous user

### LINKED ISSUE #2953: Okta Authentication - Ingestion
**Affected module**
Does it impact the UI, backend or Ingestion Framework?
Ingestion
**Describe the bug**
Okta Authentication from ingestion doesn't handle the creation and validation of access token

### PR DESCRIPTION
**Pull request by harshach**
### Describe your changes :
See #2953

#
### Type of change :

- [x] Bug fix

#
### Frontend Preview (Screenshots) :
<p align="center">For frontend related change, please link screenshots of your changes preview! Optional for backend related changes.
</p>

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [ ] I have performed a self-review of my own. 
- [ ] I have tagged my reviewers below.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#

### CODE REVIEW

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java [hunk 1]
```diff
@@ -53,4 +73,38 @@ public boolean isAdmin(AuthenticationContext ctx) {
...
+      if (user != null && (user.getIsBot() == null || !user.getIsBot())) {
+        user.setIsAdmin(true);
+      }
+      addOrUpdateUser(user);
+    } catch (EntityNotFoundException ex) {
+      User user =
+          new User()
+              .withId(UUID.randomUUID())
+              .withName(username)
+              .withEmail(username + "@domain.com")
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java:90]: Wasn't the domain specified in the config file?
If so, should that be used?

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java [hunk 2]
```diff
@@ -53,4 +73,38 @@ public boolean isAdmin(AuthenticationContext ctx) {
...
+      addOrUpdateUser(user);
+    } catch (IOException | ParseException e) {
+      LOG.error("Failed to create admin user {}", username, e);
+    }
+  }
+
+  private void addOrUpdateUser(User user) {
+    try {
+      RestUtil.PutResponse<User> addedUser = userRepository.createOrUpdate(null, user);
+      LOG.debug("Added admin user entry: {}", addedUser);
```
Comment (vivekratnavel) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java:103]:
```suggestion
      LOG.debug("Added anonymous user entry: {}", addedUser);
```

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java [hunk 3]
```diff
@@ -53,4 +73,38 @@ public boolean isAdmin(AuthenticationContext ctx) {
...
+    } catch (IOException | ParseException exception) {
+      // In HA set up the other server may have already added the user.
+      LOG.debug("Caught exception: {}", ExceptionUtils.getStackTrace(exception));
+      LOG.debug("Admin user entry: {} already exists.", user);
```
Comment (vivekratnavel) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java:107]:
```suggestion
      LOG.debug("Anonymous user entry: {} already exists.", user);
```

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java [hunk 4]
```diff
@@ -53,4 +73,38 @@ public boolean isAdmin(AuthenticationContext ctx) {
   public boolean isBot(AuthenticationContext ctx) {
     return true;
   }
+
+  private void addAnonymousUser() {
+    EntityUtil.Fields fields = new EntityUtil.Fields(FIELD_LIST, fieldsParam);
+    try {
+      User user = userRepository.getByName(null, username, fields);
+      if (user != null && (user.getIsBot() == null || !user.getIsBot())) {
+        user.setIsAdmin(true);
```
Comment (vivekratnavel) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java:82]: We don't need this for anonymous user

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java [hunk 5]
```diff
@@ -53,4 +73,38 @@ public boolean isAdmin(AuthenticationContext ctx) {
...
+      }
+      addOrUpdateUser(user);
```
Comment (vivekratnavel) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java:84]:
```suggestion
      userRepository.getByName(null, username, fields);
```

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/CatalogApplication.java [hunk 6]
```diff
@@ -198,6 +198,7 @@ private void registerAuthorizer(CatalogApplicationConfig catalogConfig, Environm
     } else {
       LOG.info("Authorizer config not set, setting noop authorizer");
       authorizer = NoopAuthorizer.class.getConstructor().newInstance();
+      authorizer.init(authorizerConf, jdbi);
```
Comment (vivekratnavel) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/CatalogApplication.java:201]: authorizerConf will always be null here.
```suggestion
      authorizer.init(null, jdbi);
```

**Review by vivekratnavel — CHANGES_REQUESTED** [review]

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java (hunk 1 above)
Comment (vivekratnavel) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/security/NoopAuthorizer.java:90]: Yes, for anonymous user, we can simply use any default domain name.

**Review by vivekratnavel — APPROVED** [review]

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/security/DefaultAuthorizer.java [hunk 7]
```diff
@@ -245,7 +245,7 @@ private User getUserFromAuthenticationContext(AuthenticationContext ctx) throws
     return userRepository.getByName(null, userName, fields);
   }
 
-  private void addOrUpdateAdmin(User user) {
+  private void addOrUpdateUser(User user) {
     try {
       RestUtil.PutResponse<User> addedUser = userRepository.createOrUpdate(null, user);
       LOG.debug("Added admin user entry: {}", addedUser);
```
[the remark this answers is not in the document]
Comment (vivekratnavel) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/security/DefaultAuthorizer.java:251]: @mithmatt I will take care of it as part of #2865

**Pull request merged**